src/server/mod.rs

```diff
@@ -1442,30 +1495,69 @@ impl LanguageServer for LspServer {
-                            },
-                        }],
-                    )])),
-                    document_changes: None,
-                    change_annotations: None,
-                };
+                if composition.exists_in(&file)
+                    && composition.resolve_with_ast(file).is_ok()
+                {
+                    let position = composition
+                        .get_stmt_position()
+                        .expect("Bad stmt position.");
+                    let range_action_item = lsp::MessageActionItem {
+                        title: LspMessageActionItem::CompositionRange
+                            .to_string(),
+                        properties: HashMap::from([(
+                            "range".to_string(),
+                            lsp::MessageActionItemProperty::Object(
+                                serde_json::to_value(HashMap::from(
+                                    [
+                                        ("start", position.start),
+                                        ("end", position.end),
+                                    ],
+                                ))
+                                .expect("Bad stmt position"),
+                            ),
+                        )]),
+                    };
+                    let params = lsp::ShowMessageRequestParams {
+                        typ: lsp::MessageType::INFO,
+                        message: LspClientCommand::AlreadyExists
+                            .to_string(),
+                        actions: Some(vec![range_action_item]),
+                    };
+                    if let Some(client) = self.get_client() {
+                        let _ = client.send_custom_notification::<ClientCommandNotification>(params).await;
+                    };
+                } else {
+                    let edit = lsp::WorkspaceEdit {
+                        changes: Some(HashMap::from([(
+                            command_params.text_document.uri.clone(),
+                            vec![lsp::TextEdit {
+                                new_text: composition.to_string(),
+                                range: {
+                                    let file =
+                                        self.store.get_ast_file(
+                                            &command_params
+                                                .text_document
+                                                .uri,
+                                        )?;
+                                    file.base.location.into()
+                                },
+                            }],
+                        )])),
+                        document_changes: None,
+                        change_annotations: None,
+                    };
+                    if let Some(client) = self.get_client() {
+                        let _ = client.apply_edit(edit, None).await;
```

Comment:
This is a case where assigning to `_` is creating a bug, the `client.apply_edit()` function returns a result and we are not checking that result for an error. Additionally the ApplyWorkspaceEditResponse is returned inside the result, which indicates if the edit was actually applied. We likely need to check that and report an error if the edit was not applied.

Reply:
Good call.
Made an update to send notice back to the UI => which can then toast the failure to update the composition. 

Note that we don't need to worry about the exact failure, nor getting out of sync with the UI, since the `did_change()` method already handles the sync. (The ActionItem sent to the UI of `LspMessageActionItem::CompositionState` does this sync.)